pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -1638,28 +1756,68 @@ impl Timeline {
             if !need_split && !last_batch {
                 return Ok(None);
             }
-            let deltas = std::mem::take(deltas);
+            let deltas: Vec<(Key, Lsn, Value)> = std::mem::take(deltas);
             if deltas.is_empty() {
                 return Ok(None);
             }
             let end_lsn = deltas.iter().map(|(_, lsn, _)| lsn).max().copied().unwrap() + 1;
+            let delta_key = PersistentLayerKey {
+                key_range: {
+                    let key_start = deltas.first().unwrap().0;
+                    let key_end = deltas.last().unwrap().0.next();
+                    key_start..key_end
+                },
+                lsn_range: lowest_retain_lsn..end_lsn,
+                is_delta: true,
+            };
+            {
+                // Hack: skip delta layer if we need to produce a layer of a same key-lsn.
```

Comment:
```suggestion
                // Hack: skip delta layer if we need to produce a layer of a same key-lsn.
                //
                // This can happen if we have removed some deltas in "the middle" of some existing layer's key-lsn-range.
                // For example, consider the case where a single delta with range [0x10,0x50) exists.
                // And we have branches at LSN 0x10, 0x20, 0x30.
                // Then we delete branch @ 0x20.
                // Bottom-most compaction may now delete the delta [0x20,0x30).
                // And that wouldnt' change the shape of the layer.
                //
                // Note that bottom-most-gc-compaction never _adds_ new data in that case, only removes.
                // That's why it's safe to skip.
```